teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,389 @@
+interface relies on HTTP status codes to indicate if processing should
+continue or not. This makes it hard to distinguish between a interceptor
+telling Triggers to stop processing and an interceptor returning an
+unexpected error (see
+[#336](https://github.com/tektoncd/triggers/issues/336)) To configure a
+webhook interceptor, a user has to setup a deployment and service and then
+configure the EventListener spec to use that service. They also have to
+manually look at the logs for both the Listener and their service to debug
+any issues.
+
+Today built in interceptors are more powerful, faster, and easier to use and
+debug. But they are not pluggable or extensible. To add a new built in
+interceptor, one would have to add it to the Triggers source tree. Webhook
+Interceptors are pluggable but aren't as powerful as built-in interceptors.
+
+The goal of this proposal is to reconcile these differences and come with a
+unified interceptor interface that supports both the pluggability of webhook
+interceptors with the advantages of the current built in interceptors
+
+### Goals
+
+1. Interceptors are pluggable i.e an operator can configure exactly which interceptors their installation of Tekton Triggers should support.
+
+2. Unify the interfaces i.e. users should be able to use and supply configuration for both built in and webhook interceptors in the same way ideally using fields in the Trigger spec.
+
+3. Provide a consistent way to represent that Trigger processing should continue without having to throw an error.
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Requirements
+
+1. Operators can install one or more built in interceptors.
+2. Operators can remove any of the built-in interceptors.
+3. Users can configure Interceptors with custom parameters including references to Secrets.
+4. Interceptor authors can add new interceptors that have the same features as any built in interceptors, namely access to the entire request and any custom user parameters
+5. Any current Interceptor configuration is still usable in the new model without requiring any manual changes on part of the end user.
+
+
+## Proposal
+
+1. Interceptors run as separate processes that communicate with the EventListener over HTTP. The EventListener will send a request to the interceptor with a body that includes incoming event details (headers, body, url), as well as any parameters needed by the interceptor (filter expression, refs to secrets etc.) The interceptor response body follows a predefined format that includes information on whether processing should continue or as well as any additional fields that the interceptor can add to the event body. This is in line with TEP-0022 which proposes a change to make the incoming event body immutable to improve auditability, and thus interceptors can only add extra fields instead of being able to modify the body or headers.
```

Comment:
yeah, and that poor wording on my part. With one interceptor serving multiple EventListeners, it is possible that one EL that has many incoming events (or has many triggers with many interceptors)  starts making too many requests to the interceptor. This could result in Listeners being bottle-necked on the interceptor processing and then failing to process their triggers (or building up a large queue of triggers/events to process).

Reply:
updated wording in the TEP.

Reply:
indeed I did :facepalm: